Expose `start_link_opts_schema`

jvf opened the pull request:
As a maintainer of a library that uses Xandra and NimbleOptions and passes options through to `Xandra.start_link/` I am interested in documenting the options directly inside my code. For this it would be helpful to be able to access the `start_link_opts_schema`.

jvf commented:
I was mainly planning to use this for documenting the options.

I am also thinking of moving my library to NimbleOptions. For this it might be useful to validate the options as early as possible. Also NimbleOptions would complain about additional options (could be circumvented of course, but for this it would be nice to have at least the option keys without copying). Error handling would become easier I think if I do not need to catch validation errors from Xandra.

> This schema is not "public API" and we might change its internal workings without notice.

I am paying pretty close attention to Xandra so this would not be a problem for me. I understand the concern though. That is why put this as `@doc false`, my idea was to signal that this is more of a use-at-your-own risk "private" API.

whatyouhide commented:
Hey @jvf, thanks for the patience. I gave this a bit more thought. There are good chance that this will be refactored, moved, changed, and so on, so I still feel a bit uncomfortable to expose it as a "private API" (I don't think there's necessarily such a thing as a *private API* that can be *exposed* 😅).

Let me pitch you one last idea. In your library, can you:
1. Get a bunch of options from the user
2. Take away **your** options with something like
    ```elixir
    {my_opts, xandra_opts} = Keyword.split(opts, [:my_opt1, :my_opt2, :...])
    ```
3. Pass the `xandra_opts` to Xandra, which will blow up nicely if something is wrong
4. For documentation, say something like "see the Xandra options" and add a link for that

This is how I'd likely do it if I were doing something like what I think you're doing 😄

jvf commented:
Hi @whatyouhide, thank you for your suggestions! If can't get the options schema these are good ideas to work around it. 

I still think it would be useful to be able to work programmatically with the options of a library I use. Some examples:

Imagine a service app (i.e. an app that will constitute its own release, implements the `Application` behaviour and is configured through application environment). To reduce the problems associated with application environment configuration I like to read the complete application environment once in the top level supervisor and pass around all the options to all the children. Somewhere in the supervision hierarchy Xandra is started. So I would like to pull out all Xandra related options. For this I need a list of Xandra options keys that I think would be best obtained directly from Xandra. (Sure, I could put all the options under a sub-key. But I need some of the options somewhere else as well, so now I would need either duplicate the options or reach into the xandra sub-key, which is also not so nice).

For testing, this could also be nice. I imagine something like having a list of option keys that I use in my app and test it against the option keys obtained from the Xandra options schema. If I do this in a CI pipeline that runs against Xandra master I could have an early warning in CI if something changes.

For documentation: I am familiar with the "see the XY options" way of doing things and I know it is used by a lot of libraries. But I think the user experience of requiring a user to jump from codebase to codebase to check the documentation is not optional and could be improved. A user of my library would need to jump from my docs to Xandra and then to DBConnection in some cases. Having the documentation in one place would IMHO improve the user experience. In Xandra you also do this manually for example for `pool_size` (which is passed down to DBConenction as-is if I read the code correctly). My hope was that if many apps use NimbleOptions exposing the options schema could be a way of improving the user experience.

> There are good chance that this will be refactored, moved, changed, and so on, so I still feel a bit uncomfortable to expose it

I am also trying to understand what kind of problematic changes you anticipate. If the options themselves change this would be a change of the public API anyway and users of Xandra would need to deal with it accordingly anyway, right? Or do anticipate the format of NimbleOptions changing? Or stopping to use NimbleOptions?

whatyouhide commented:
@jvf okay, after some thought, I'll go ahead and merge this. However, this will indeed be **private API**. This means that I might refactor Xandra to not use NimbleOptions, I might refactor the schema, I might rename this function, or any of the things I usually do to private APIs without notice or mentions in the CHANGELOG. It will definitely be "use at your own risk" 😄 Hope that works for you, and thanks for the discussion! 💟

jvf commented:
Awesome, much appreciated. 💜
